### ISSUE TITLE
File-watchers do not trigger (when used with CorgiEngine)

### ISSUE THREAD

**Issue by WuKaiYi** [handzlikchris/FastScriptReload#93]
After I upgraded from 2022.2 to 2022.3 the automatic Script Reload did not work without any prompts or errors

**Comment by D0rkKnight** [comment]
^ bump, am experiencing the same issues on Unity 2022.3.2f1. The UI windows work but my scripts don't get reloaded.

**Comment by handzlikchris** [comment]
Can you provide some more details please.

When you run demo scene according to instructions are you getting any
errors?

**Comment by D0rkKnight** [comment]
It's weird
I ran the test scene according to the instructions on a new project and it worked perfectly
But on my current game project with a lot of files and whatnot nothing is printed to the console when I edit a script in playmode (so no errors)

I tried turning on debug mode and it threw a compile error saying LoggerScoped wasn't recognized in ProjectTypeCache.cs 
<img width="546" alt="image" src="https://github.com/handzlikchris/FastScriptReload/assets/20606858/40680837-f578-420a-a03f-50089fc0860e">

I commented the line out and it compiled, but when running the test scene by the instructions I still don't get any console output aside from the manager being created
It just looks like this:
<img width="469" alt="image" src="https://github.com/handzlikchris/FastScriptReload/assets/20606858/2cbfe112-1d46-49ef-96cb-1f22269c90a3">

**Comment by handzlikchris** [comment]
That is indeed quite weird.

If there are no errors I'm thinking perhaps file watcher didn't detect the
change for some reason.

Are you on windows? What code editor are you using?

**Comment by WuKaiYi** [comment]
No error, warning, or even FSR can be initialized.

FSR: Created Manager
UnityEngine.Debug:Log (object)
ImmersiveVrToolsCommon.Runtime.Logging.LoggerScoped:LogDebug (string)

**Comment by D0rkKnight** [comment]
Yep, am on Windows 11. I use VSCode mainly but it doesn't work with Visual Studio either after a bit of testing.

VSCode: v1.80.0
Visual Studio: Community 2022 v17.6.4

**Comment by handzlikchris** [comment]
Tested on 2022.3.4 - all ok.

If there's no logs I think it'll be down to File Watchers, do you default config there?

![image](https://github.com/handzlikchris/FastScriptReload/assets/6343986/847977c8-0b7a-47c7-935e-16b465ac20b5)

You can also add breakpoint here: `FastScriptReloadManager.OnWatchedFileChange` - and see if making changes even get's the code execution there.

**Comment by WuKaiYi** [comment]
> Tested on 2022.3.4 - all ok.
> 
> If there's no logs I think it'll be down to File Watchers, do you default config there?
> 
> ![image](https://user-images.githubusercontent.com/6343986/253733888-847977c8-0b7a-47c7-935e-16b465ac20b5.png)
> 
> You can also add breakpoint here: - and see if making changes even get's the code execution there.`FastScriptReloadManager.OnWatchedFileChange`

I tried to add breakpoints, did not execute,File Watchers I kept the same Settings

**Comment by WuKaiYi** [comment]
I solved it and now it can run perfectly on 2022.3.4. I will first return the version to 2021 and clear the error caused by the version change before upgrading to 2022.3.4. By contrast
Packages manifest.json
Packageslpackages-lock.json
![image](https://github.com/handzlikchris/FastScriptReload/assets/34147620/140e8405-96d1-4982-9e06-e23c7bdb28c0)
I found that it could be "com.unite.feature.2d ": "2.0.0", which is the cause

**Comment by D0rkKnight** [comment]
Can confirm that OnWatchedFileChange is not triggering. I placed a Debug.Log there and nothing gets logged.

My project has a lot of files, and when I delete all the assets except for FSR (after backing up, ofc), the file watcher begins working again.
This leads me to think that the file watchers may be struggling with the number of files present.

**Comment by handzlikchris** [comment]
Alrigh, I've seen those file watchers having some issues. With not firing
at all - is there any thing else in you proj that uses file watchers?

You can also try to narrow down which folder is watched in fsr settings.
But if they never fire it's probably something else

**Comment by D0rkKnight** [comment]
Nah, just FSR is using file watchers afaik. I had it watch some smaller folders but still the file watchers weren't being triggered.

I've however localized the issue to something with CorgiEngine. Creating a fresh project, FSR works until I import CorgiEngine.
My understanding is the issue may be from 1 of 3 places
1. the imported CorgiEngine assets
2. the imported project settings
3. some legacy Unity package that is being imported as a dependency

**Comment by D0rkKnight** [comment]
To follow up, the issue has been isolated to the MMTools library which comes packaged with CorgiEngine.

**Comment by handzlikchris** [comment]
Thanks for narrowing it down.

@reunono - I wonder if you can help, is CorgiEngine using FileWatchers? I've some issues with Unity when sometimes if multiple file watchers are created just one will receive callback.

**Comment by D0rkKnight** [comment]
Ah sorry, I may have been mistaken. Some further testing suggests that the filewatchers might just be really slow.

With the full CorgiEngine (this includes MMData), a hot reload took 18 minutes.
Meanwhile, with CorgiEngine trimmed down to only MMData's base source code, the hot reload takes only a couple seconds.
It seems that depending on how much source code I leave around, the filewatchers respond faster or slower.

For reference, CorgiEngine has 1099 .cs files in its source including its packaged dependencies (MMTools, MMInterface, MMFeedbacks, MMInventory). 
But I also can't replicate this by generating 1099 .cs files each containing a MonoBehaviour. So maybe it's to do with the quantity of text in the files or it's something else entirely.

Also, there is no mention of FileSystemWatcher or FileWatcher in the CorgiEngine source code that is generating the issue.

**Comment by reunono** [comment]
Hey there! No, no use of FileWatchers in the engine (or its associated libraries).
I can confirm however that it is a heavy codebase, lots of files, the vast majority being regular MonoBehaviours, pure gameplay code, a few editor scripts, lots of sprites/3D models/textures to power the demos, and a few shaders. 
Let me know if I can help solve this!

**Comment by handzlikchris** [comment]
Thanks for confirming Reunono - looks like file watchers are just being slow with this many files.

@D0rkKnight You can narrow down file watchers scope - but I think we're already tried that.

On some Unity versions file watchers are nearly completely broken (they don't report full path but just file-name on change).

As a nasty-workaround FSR then tries to do a root directory search for that file and picks first with that name. This could be happening, do you see 'Fast Script Reload - Unity File Path Bug - Warning!' anywhere in your logs?

More info on Unity issue:
https://issuetracker.unity3d.com/issues/filesystemwatcher-returns-bad-file-path

**Comment by D0rkKnight** [comment]
No, I don't see such a warning anywhere. I also put in some logging and it looks like the slowdown is between the file edit and the start of OnWatchedFileChange. Since the File Path Bug warning is in the body of OnWatchedFileChange we can hopefully rule that out.

**Comment by reunono** [comment]
Chris, would it help you reproduce and fix this if I provided you with a free copy of the engine? 
Or is it more of a general issue and it won't help much?

**Comment by handzlikchris** [comment]
Thanks that's really kind offer. I acutally do have a copy, your tools are quite popular :)!

I don't think it's that though, likely is just amount of files that makes the issue more visible. Some other users also mention similar issues with FileWatchers from time to time.

**Comment by reunono** [comment]
Alright! I wish I could help but I have almost zero FileWatchers experience :)

**Comment by handzlikchris** [comment]
yeah - that'd be fantastic.

If I could suggest few things which will make it easier to merge into FSR once done:

- there's an existing data structure that's persisted for user: FileWatcherSetupEntry - would be good to use that - this way people could set their watchers on FileWatchers options page and if something doesn't work right they'll be able to tick the box for 'custom File Watching' 

- all watching logic could be done separately from main FSR workflow, I'd imagine it'll run in separate thread and check watched files/folder. When there's a change in file all that's needed:
- is adding an entry to
`FastScriptReload.Editor.FastScriptReloadManager._dynamicFileHotReloadStateEntries` (might need to expose a method that'll allow to add there)
> you can also check out how existing FileWatche implementation adds that in `FastScriptReload.Editor.FastScriptReloadManager.OnWatchedFileChange`
- FSR will pick up changes normally

> All that should really limit what you need to change in FSR and there'll be little place for errors to creep in.

As for options/flag I can get that added later, if you want to have a look they are all defined in `FastScriptReload.Editor.FastScriptReloadPreference`

defining those is straight forward, eg
```
        public static readonly ToggleProjectEditorPreferenceDefinition EnableExperimentalEditorHotReloadSupport = new ToggleProjectEditorPreferenceDefinition(
            "(Experimental) Enable Hot-Reload outside of play mode", "EnableExperimentalEditorHotReloadSupport", false);
```

And then to render on settings page it's just the matter of going to section definition around here
```
new GuiSection("Advanced", new List<ClickableElement>
{
    new ChangeMainViewButton("File Watchers", (screen) => {
   //some other code already here
   //add like so
   ProductPreferenceBase.RenderGuiAndPersistInput(FastScriptReloadPreference.FileWatcherSetupEntries);
}
```
And that'll handle persistence / rendering to user etc

Feel free to drop me a message if you need any info.

Good luck!

**Comment by D0rkKnight** [comment]
Got it, thanks for the info! Will be sure to ask any questions if issues crop up.

**Comment by D0rkKnight** [comment]
Does _dynamicFileHotReloadStateEntries track newly created files and deleted files as well? 

Also would it be better to contact you on Github or on Discord? @handzlikchris

**Comment by handzlikchris** [comment]
new and deleted files are generally not handled at all. FSR won't be able to reload them.

At core it 'redirects' your existing calls at runtime using modding framework Harmony.
There's some 'magic' that allows you to add new stuff at runtime but really that changes your code to use some dictionaries for new fields and such.

But mostly it'll work with changes to existing code only.

Deleting stuff is not really supported, if you want method to stop doing whatever is doing then early return is best option.

**Comment by D0rkKnight** [comment]
Alright, I've got the file watching working on my end. The code is all manners of scuffed (and I'm not sure if I did the threading right) but I'm working on getting the code onto a fork for review.

With CorgiEngine (a beefy codebase) we have around 300ms latency on the 2nd thread.
I put an option on the file watcher options page to toggle whether or not to use the custom file watcher. If you use the custom filewatcher the unity default filewatcher is disabled.
Right now I'm not super clear on the application life cycle so I have the custom filewatcher initializing its file-hash table when file-watchers would normally be deployed (at StartWatchingDirectoryAndSubdirectories)
And I have the file watcher checking the filesystem every second perpetually, from the moment the Editor loads. Ideally this should be bound to only play mode unless reloading is enabled in editor mode.

**Comment by D0rkKnight** [comment]
Alright the fork has been created
https://github.com/D0rkKnight/FastScriptReload

Here's the commit with all the changes https://github.com/handzlikchris/FastScriptReload/commit/b463817fa3ab0a68ddf2b438e49d892b029c78ee

**Comment by handzlikchris** [comment]
@D0rkKnight nice one, just had a quick glance - looks good, thanks!

Btw. I wouldn't want to steal your work, if you create PR I'll merge it in to feature branch, do some testing and merge into main branch once ready. 

This way your contribution will be recorded.

**Comment by D0rkKnight** [comment]
Sounds good, I'll clean the code up and submit the pull request!

**Comment by BoomerSchmirtz** [comment]
Just following this issue :)

**Comment by D0rkKnight** [comment]
Alright, @handzlikchris the pull request is in. I'll let you take it from here :)

**Comment by handzlikchris** [comment]
closed with #100
